v2/spanner-change-streams-to-sharded-file-sink/src/test/java/com/google/cloud/teleport/v2/templates/SpannerChangeStreamToGcsMultiShardIT.java

```diff
@@ -0,0 +1,224 @@
+  private static PipelineLauncher.LaunchInfo jobInfo;
+  private static String spannerDatabaseName = "";
+  private static String spannerMetadataDatabaseName = "";
+  private static GcsResourceManager gcsResourceManager;
+
+  /**
+   * Does the following setup:
+   *
+   * <p>1. Creates a Spanner database with a given table 2. Creates a shard file with the connection
+   * details 3. The session file for the same is taken from the resources and uploaded to GCS 4.
+   * Places the session file and shard file in GCS 5. Creates the change stream in Spanner database
+   * 6. Creates the metadata database 7. Launches the job to read from Spanner and write to GCS
+   *
+   * @throws IOException
+   */
+  @Before
+  public void setUp() throws IOException {
+    skipBaseCleanup = true;
+    synchronized (SpannerChangeStreamToGcsMultiShardIT.class) {
+      testInstances.add(this);
+      if (jobInfo == null) {
+        gcsResourceManager = createGcsResourceManager(getClass().getSimpleName());
+        spannerResourceManager = createSpannerResourceManager();
+        spannerMetadataResourceManager = createSpannerMetadataResourceManager();
+        prepareLaunchParameters(
+            gcsResourceManager,
+            spannerResourceManager,
+            spannerMetadataResourceManager,
+            spannerDdl,
+            sessionFileResourceName);
+        createAndUploadShardConfigToGcs();
+        jobInfo =
+            launchReaderDataflowJob(
+                gcsResourceManager,
+                spannerResourceManager,
+                spannerMetadataResourceManager,
+                getClass().getSimpleName(),
+                null,
+                null,
+                true);
+      }
+    }
+  }
+
+  @AfterClass
+  public static void cleanUp() throws IOException {
+    for (SpannerChangeStreamToGcsMultiShardIT instance : testInstances) {
+      instance.tearDownBase();
+    }
+    ResourceManagerUtils.cleanResources(
+        spannerResourceManager, spannerMetadataResourceManager, gcsResourceManager);
+  }
+
+  private void createAndUploadShardConfigToGcs() throws IOException {
+    List<String> shardNames = new ArrayList<>();
+    shardNames.add("testShardB");
+    shardNames.add("testShardC");
+    JsonArray ja = new JsonArray();
+
+    for (String shardName : shardNames) {
+      Shard shard = new Shard();
+      shard.setLogicalShardId(shardName);
+      shard.setUser("dummy");
+      shard.setHost("dummy");
+      shard.setPassword("dummy");
+      shard.setPort("3306");
+      JsonObject jsObj = (JsonObject) new Gson().toJsonTree(shard).getAsJsonObject();
+      ja.add(jsObj);
+    }
+
+    String shardFileContents = ja.toString();
+    LOG.info("Shard file contents: {}", shardFileContents);
+    // -DartifactBucket has the bucket name
+    gcsResourceManager.createArtifact("input/shard.json", shardFileContents);
+  }
+
+  private void writeSpannerDataForSingers(int singerId, String firstName, String shardId) {
```

Comment:
We would need one more test case - even if there are multiple shards, Spanner writes are happening to one shard only.

Reply:
Can be marked as TODO

Reply:
Spanner writes are happening for two shards : testShardB and testShardC. Am I missing something here ?

Reply:
I meant that shard file has two shards, while the writes are happening only on one shard. But it's not a P0, resolving the comment.